distributed/core.py

```diff
@@ -877,42 +884,42 @@ def __call__(self, addr=None, ip=None, port=None):
             comm._pool = weakref.ref(self)
             self._created.add(comm)
         except Exception:
+            self.semaphore.release()
             raise
-        occupied.add(comm)
+        finally:
+            self._n_connecting -= 1
 
-        if self.open >= self.limit:
-            self.event.clear()
+        occupied.add(comm)
 
         return comm
 
     def reuse(self, addr, comm):
         """
         Reuse an open communication to the given address.  For internal use.
         """
-        try:
-            self.occupied[addr].remove(comm)
-        except KeyError:
-            pass
```

Comment:
I looked for the call sites of `ConnectionPool.reuse`; fortunately, there were only two, and in both cases, `comm` refers to an object managed by the pool which should thus not lead to a `KeyError` here.
Note, though, that I did comment out one line of tests here:
https://github.com/dask/distributed/pull/3005/commits/fb30744692c309ccd145de34fe21d39241c07f44#diff-863a91ce704c12168c86ab5e8a99ede0L622

Reply:
Looking at the tests, I stand corrected: it seems to be important after all (`test_gather_then_submit_after_failed_workers` failed because of this). I'll prepare a fix.